tests/unit/s2n_fork_generation_number_test.c

```diff
@@ -189,6 +190,9 @@ static int s2n_unit_test_clone_child_process(void *parent_process_fgn)
     /* Verify in threads */
     EXPECT_EQUAL(s2n_unit_test_thread(return_fork_generation_number), S2N_SUCCESS);
 
+    /* Clone can't invoke atexit. Call rand cleanup to close /dev/urandom */
+    POSIX_GUARD_RESULT(s2n_rand_cleanup());
```

Comment:
Fun :( Did you find documentation of that behavior?

Reply:
I first noticed this by my observation. After the cloned child process exit, atexit cleanup (with no callbacks reset) is not fully executed, which leads to leaking dev urandom. This observation is reflected in this [comment](https://github.com/aws/s2n-tls/pull/4835#discussion_r1802028152).

I did some further investigations by reading `atexit` [man page](https://man7.org/linux/man-pages/man3/atexit.3.html), which says
> The atexit() function registers the given function to be called
       at normal process termination, either via [exit(3)](https://man7.org/linux/man-pages/man3/exit.3.html) or via return
       from the program's main().

The current set up in `s2n_fork_number_generation_test.c` uses return a success code for the process to terminates, which doesn't meet the requirements to trigger `atexit` (it's neither via exit, nor returns in main).
https://github.com/aws/s2n-tls/blob/9cd117f3d9a94912d5e8f2a61ffbfd31f11d2946/tests/unit/s2n_fork_generation_number_test.c#L192-L193.

The exit behavior of `clone` is specified in the [man page](https://man7.org/linux/man-pages/man2/clone.2.html) specifically

> When the fn(arg) function returns, the child process terminates.
       The integer returned by fn is the exit status for the child
       process.  The child process may also terminate explicitly by
       calling [exit(2)](https://man7.org/linux/man-pages/man2/exit.2.html) or after receiving a fatal signal.

Therefore, I think a better and more conventional solution to invoke our atexit clean up is to do `exit(EXIT_SUCCESS)`, instead of `return EXIT_SUCCESS`. I have implemented locally, and tested it. This will also fix the leaking dev urandom problem.
https://github.com/boquan-fang/s2n-tls/blob/2036a245e3e1bcbef6ea9c8ddfc9595f76b29b55/tests/unit/s2n_fork_generation_number_test.c#L177-L195

I think directly call exit for cloned process is a better solution than unconventionally calling `s2n_rand_cleanup`. Let me know what you think.

Reply:
If this can be fixed by just directly calling exit, that does sound like a better solution to me.